### ISSUE TITLE
UndocumentedPublicProperty false positive when property is documented in class header

### ISSUE THREAD

**Issue by evg64** [detekt/detekt#5114]
I have a problem with how rule UndocumentedPublicProperty works.
Say, we have the following class:
```
/**
 * Class description
 *
 * @property bar Property description
 */
class Foo {
    val bar = 2
}
```

## Expected Behavior
Detekt says there are no problems with documentation (KDoc) of property `bar`.

## Observed Behavior
Deteks says, that property `bar` has no documentation.

## Steps to Reproduce
Run detekt over the class above.

## Context
KDoc supports documenting properties in class header. Detekt does not support that. It only expects property`s documentation in declaration-site.
Maybe it is worth adding boolean-option to rule named UndocumentedPublicProperty. The option would determine, whether the rule should expect KDoc not only in declaration-site, but also in class header.

**Comment by marschwar** [comment]
To me the [documentation](https://kotlinlang.org/docs/kotlin-doc.html#property-name) seems to suggest that this is reserved for properties that are defined in the primary constructor. I have not checked if that is supported by detekt or not.

**Comment by schalkms** [comment]
This is not a bug, in my opinion. The detekt rule `UndocumentedPublicProperty` doesn't diverge from the KDoc specification. As seen in the [doc](https://kotlinlang.org/docs/kotlin-doc.html#property-name), `@property` should only be used for documenting parameters declared in the primary constructor.

> **@property name﻿**
Documents the property of a class which has the specified name. This tag can be used for documenting properties declared in the primary constructor, where putting a doc comment directly before the property definition would be awkward.

If you want to diverge from the official KDoc style guide, feel free to copy the _native_ detekt rule `UndocumentedPublicProperty` to a [3rd party rule](https://detekt.dev/docs/introduction/extensions).

**Comment by evg64** [comment]
1)
> @property should only be used for documenting parameters declared in the primary constructor

Please, read the documentation carefully. Nothing says `@property` should ONLY be used for such a purpose. Insdead, documentation says, `@property` CAN be used for documenting properties in constructor.

2) Back to the following class:
```
/**
 * Class description
 *
 * @property bar Property description
 */
class Foo {
    val bar = 2
}
```
Use [dokka](https://github.com/Kotlin/dokka) to generate documentation for it:
```
./gradlew dokkaHtml
```

What you will see:

![Screenshot 2022-07-25 at 15 39 07](https://user-images.githubusercontent.com/8549356/180779309-475616c2-1854-4394-abd3-d2ea6786e634.png)

It means, that dokka supports working with properties` documentation in class header.
What is more, IDE (intellij idea) supports it as well:

<img width="393" alt="Screenshot 2022-07-25 at 15 41 01" src="https://user-images.githubusercontent.com/8549356/180779686-7d6c0115-733f-4747-9f52-d0641a6da7d9.png">


To sum up, 1) tools work quite ok with documenting properties in class body, 2) documentation does not prohibit doing that. That\`s why I would not say documenting properties in class body is strictly incorrect. I believe, it is worth adding option to `UndocumentedPublicProperty` which would help to switch between two behaviours: whether or not to take into account documentation in class body.

P.S. I can support this as detekt extension, but I believe, it would be better, if detekt core supported it.
P.P.S. Closing issues without waiting for author\`s answer is not the most polite thing possible. It is like saying "I`m not going to modify my code no matter what you say".

**Comment by schalkms** [comment]
Thanks for providing a sample!
I have reopened the issue. No worries.

It’s better suited in the rule itself. The change should happen in **UndocumentedPublicProperty**.

**Comment by evg64** [comment]
It looks like a comment is required to prevent this issue from closing.
